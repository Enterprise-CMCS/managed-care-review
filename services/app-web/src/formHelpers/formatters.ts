import { dayjs } from '@mc-review/dates'
import { FileItemT } from '../components'
import { GenericDocument, ActuaryContact } from '../gen/gqlClient'
import { S3ClientT } from '../s3'
import { v4 as uuidv4 } from 'uuid'

// Convert form fields to pass data to api. GQL handles null for empty fields.
const formatForApi = (attribute: string): string | null => {
    if (attribute === '') {
        return null
    }
    return attribute
}

// Convert api data for use in form.  Form fields must be a string.
// Empty values as an empty string, dates in date picker as YYYY-MM-DD, boolean as "Yes" "No" values
function formatForForm<T>(attribute: T): string {
    if (attribute === null || attribute === undefined) {
        return ''
    } else if (attribute instanceof Date) {
        return dayjs(attribute).utc().format('YYYY-MM-DD')
    } else if (typeof attribute === 'boolean') {
        return attribute ? 'YES' : 'NO'
    } else {
        return attribute.toString()
    }
}

// This function can be cleaned up when we move off domain types and only use graphql
const formatActuaryContactsForForm = (
    actuaryContacts?: ActuaryContact[]
): ActuaryContact[] => {
    return actuaryContacts && actuaryContacts.length > 0
        ? actuaryContacts.map((contact) => {
              const {
                  name,
                  titleRole,
                  email,
                  actuarialFirm,
                  actuarialFirmOther,
              } = contact
              return {
                  name: name ?? '',
                  titleRole: titleRole ?? '',
                  email: email ?? '',
                  actuarialFirmOther: actuarialFirmOther ?? undefined,
                  actuarialFirm: actuarialFirm ?? undefined,
              }
          })
        : [
              {
                  name: '',
                  titleRole: '',
                  email: '',
                  actuarialFirm: undefined,
                  actuarialFirmOther: '',
              },
          ]
}

const formatAddtlActuaryContactsForForm = (
    actuaryContacts?: ActuaryContact[]
): ActuaryContact[] => {
    return actuaryContacts && actuaryContacts.length > 0
        ? actuaryContacts.map((contact) => {
              const {
                  name,
                  titleRole,
                  email,
                  actuarialFirm,
                  actuarialFirmOther,
              } = contact
              return {
                  name: name ?? '',
                  titleRole: titleRole ?? '',
                  email: email ?? '',
                  actuarialFirmOther: actuarialFirmOther ?? undefined,
                  actuarialFirm: actuarialFirm ?? undefined,
              }
          })
        : []
}

const formatFormDateForGQL = (attribute: string): string | undefined => {
    return attribute === '' ? undefined : attribute
}

const formatDocumentsForGQL = (fileItems: FileItemT[]): GenericDocument[] => {
    return fileItems.reduce((cleanedFileItems, fileItem) => {
        if (fileItem.status === 'UPLOAD_ERROR') {
            console.info(
                'Attempting to save files that failed upload, discarding invalid files'
            )
        } else if (fileItem.status === 'SCANNING_ERROR') {
            console.info(
                'Attempting to save files that failed scanning, discarding invalid files'
            )
        } else if (fileItem.status === 'DUPLICATE_NAME_ERROR') {
            console.info(
                'Attempting to save files that are duplicate names, discarding duplicate'
            )
        } else if (!fileItem.s3URL) {
            console.info(
                'Attempting to save a seemingly valid file item is not yet uploaded to S3, this should not happen on form submit. Discarding file.'
            )
        } else if (!fileItem.sha256) {
            console.info(
                'Attempting to save a seemingly valid file item does not have a sha256 yet. this should not happen on form submit. Discarding file.'
            )
        } else {
            cleanedFileItems.push({
                name: fileItem.name,
                s3URL: fileItem.s3URL,
                sha256: fileItem.sha256,
                dateAdded: fileItem.dateAdded ?? undefined,
            })
        }
        return cleanedFileItems
    }, [] as GenericDocument[])
}

const formatDocumentsForForm = ({
    documents,
    getKey,
}: {
    documents?: GenericDocument[]
    getKey: S3ClientT['getKey'] // S3 function to call when formatting to double check we have valid documents, probably the backend should be doing this to reduce client async errors handling with bad data
}): FileItemT[] => {
    if (!documents) return []

    return (
        documents.map((doc) => {
            const key = getKey(doc.s3URL)
            if (!key) {
                // If there is no key, this means the file saved on a submission cannot be parsed or does not exist on s3.
                // We still include the file in the list displayed to the user, but with an error.
                return {
                    id: uuidv4(),
                    name: doc.name,
                    key: 'INVALID_KEY',
                    s3URL: undefined,
                    sha256: doc.sha256,
                    status: 'UPLOAD_ERROR',
                    dateAdded: doc.dateAdded,
                }
            }
            return {
                id: uuidv4(),
                name: doc.name,
                key: key,
                s3URL: doc.s3URL,
                sha256: doc.sha256,
                status: 'UPLOAD_COMPLETE',
                dateAdded: doc.dateAdded,
            }
        }) || []
    )
}

// DEPRECATED
// Domain helpers are for HPP code. We are migrating off this in favor of directly using GQL utilities

const formatFormDateForDomain = (attribute: string): Date | undefined => {
    if (attribute === '') {
        return undefined
    }
    return dayjs.utc(attribute).toDate()
}

const formatDocumentsForDomain = (
    fileItems: FileItemT[]
): GenericDocument[] => {
    return fileItems.reduce((cleanedFileItems, fileItem) => {
        if (fileItem.status === 'UPLOAD_ERROR') {
            console.info(
                'Attempting to save files that failed upload, discarding invalid files'
            )
        } else if (fileItem.status === 'SCANNING_ERROR') {
            console.info(
                'Attempting to save files that failed scanning, discarding invalid files'
            )
        } else if (fileItem.status === 'DUPLICATE_NAME_ERROR') {
            console.info(
                'Attempting to save files that are duplicate names, discarding duplicate'
            )
        } else if (!fileItem.s3URL) {
            console.info(
                'Attempting to save a seemingly valid file item is not yet uploaded to S3, this should not happen on form submit. Discarding file.'
            )
        } else if (!fileItem.sha256) {
            console.info(
                'Attempting to save a seemingly valid file item does not have a sha256 yet. this should not happen on form submit. Discarding file.'
            )
        } else {
            cleanedFileItems.push({
                name: fileItem.name,
                s3URL: fileItem.s3URL,
                sha256: fileItem.sha256,
            })
        }
        return cleanedFileItems
    }, [] as GenericDocument[])
}

const formatYesNoForProto = (
    attribute: string | undefined
): boolean | undefined => {
    if (attribute === 'YES') {
        return true
    }
    if (attribute === 'NO') {
        return false
    }
    return undefined
}

export {
    formatForApi,
    formatForForm,
    formatYesNoForProto,
    formatFormDateForDomain,
    formatDocumentsForDomain,
    formatDocumentsForForm,
    formatActuaryContactsForForm,
    formatAddtlActuaryContactsForForm,
    formatDocumentsForGQL,
    formatFormDateForGQL,
}
