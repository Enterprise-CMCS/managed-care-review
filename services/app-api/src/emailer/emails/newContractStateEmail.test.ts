import {
    testEmailConfig,
    mockContract,
    mockMNState,
} from '../../testHelpers/emailerHelpers'
import {
    formatContractSubmissionType,
    packageName,
} from '@mc-review/submissions'
import { newContractStateEmail } from './index'
import { formatEmailAddresses } from '../formatters'
import type { ContractType, RateFormDataType } from '../../domain-models'

const defaultSubmitters = ['submitter1@example.com', 'submitter2@example.com']

test('to addresses list includes submitter emails', async () => {
    const sub = mockContract()
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining(defaultSubmitters),
        })
    )
})

test('to addresses list includes all state contacts on submission', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.stateContacts = [
        {
            name: 'test1',
            titleRole: 'Foo1',
            email: 'test1@example.com',
        },
        {
            name: 'test2',
            titleRole: 'Foo2',
            email: 'test2@example.com',
        },
    ]
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    sub.packageSubmissions[0].contractRevision.formData.stateContacts.forEach(
        (contact) => {
            expect(template).toEqual(
                expect.objectContaining({
                    toAddresses: expect.arrayContaining([contact.email]),
                })
            )
        }
    )
})

test('to addresses list does not include duplicate state receiver emails on submission', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.stateContacts = [
        {
            name: 'test1',
            titleRole: 'Foo1',
            email: 'test1@example.com',
        },
        {
            name: 'test1',
            titleRole: 'Foo1',
            email: 'test1@example.com',
        },
    ]
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template.toAddresses).toEqual([
        'test1@example.com',
        ...defaultSubmitters,
        ...testEmailConfig().devReviewTeamEmails,
    ])
})

test('subject line is correct and clearly states submission is complete', async () => {
    const sub = mockContract()
    const defaultStatePrograms = mockMNState().programs
    const name = packageName(
        sub.stateCode,
        sub.stateNumber,
        sub.packageSubmissions[0].contractRevision.formData.programIDs,
        defaultStatePrograms
    )

    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `${name} was successfully submitted.`
            ),
        })
    )
})

test('includes mcog, rate, and team email addresses', async () => {
    const sub = mockContract()
    const defaultStatePrograms = mockMNState().programs
    const name = packageName(
        sub.stateCode,
        sub.stateNumber,
        sub.packageSubmissions[0].contractRevision.formData.programIDs,
        defaultStatePrograms
    )

    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to mcog@example.com`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to rates@example.com`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to ${formatEmailAddresses(
                    testEmailConfig().helpDeskEmail
                )}`
            ),
        })
    )
})

test('includes link to submission', async () => {
    const sub = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractType =
        'AMENDMENT'
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                `http://localhost/submissions/${formatContractSubmissionType(sub.contractSubmissionType)}/${sub.id}`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyHTML: expect.stringContaining(
                `href="http://localhost/submissions/${formatContractSubmissionType(sub.contractSubmissionType)}/${sub.id}"`
            ),
        })
    )
})

test('includes information about what is next', async () => {
    const sub = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractType =
        'AMENDMENT'
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining('What comes next:'),
        })
    )
})

test('does not include information about what is next for CHIP-only submissions', async () => {
    const sub = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractType =
        'AMENDMENT'
    sub.packageSubmissions[0].contractRevision.formData.populationCovered =
        'CHIP'
    const defaultStatePrograms = mockMNState().programs
    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template.bodyText).not.toContain('What comes next:')
    expect(template.bodyText).not.toContain('Check for completeness:')
    expect(template.bodyText).not.toContain('CMS review:')
    expect(template.bodyText).not.toContain('Questions:')
    expect(template.bodyText).not.toContain('Decision:')
})

test('includes expected data summary for a contract and rates submission State email', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractDateStart =
        new Date('01/01/2021')
    sub.packageSubmissions[0].contractRevision.formData.contractDateEnd =
        new Date('01/01/2025')
    sub.packageSubmissions[0].contractRevision.formData.contractType = 'BASE'
    sub.packageSubmissions[0].rateRevisions[0].formData = {
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('01/02/2021'),
        rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
        rateCertificationName:
            'MCR-MN-0003-MSHO-RATE-20210101-20220101-CERTIFICATION-20210102',
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'MERCER',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    const defaultStatePrograms = mockMNState().programs

    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Submission type: Contract action and rate certification'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Rating period: 01/01/2021 to 01/01/2022'
            ),
        })
    )

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Contract effective dates: 01/01/2021 to 01/01/2025'
            ),
        })
    )
    //Expect only have 1 rate names using regex to match name pattern specific to rate names.
    expect(
        template.bodyText?.match(
            /-RATE-[\d]{8}-[\d]{8}-(?:CERTIFICATION|AMENDMENT)-[\d]{8}/g
        )?.length
    ).toBe(1)
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.packageSubmissions[0].rateRevisions[0].formData
                    .rateCertificationName!
            ),
        })
    )
})

test('includes expected data summary for a multi-rate contract and rates submission State email', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractDateStart =
        new Date('01/01/2021')
    sub.packageSubmissions[0].contractRevision.formData.contractDateEnd =
        new Date('01/01/2025')
    sub.packageSubmissions[0].contractRevision.formData.contractType = 'BASE'
    const rate1FormData: RateFormDataType = {
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('01/02/2021'),
        rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
        rateCertificationName:
            'MCR-MN-0003-MSHO-RATE-20210101-20220101-CERTIFICATION-20210102',
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'MERCER',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    const rate2FormData: RateFormDataType = {
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('02/02/2022'),
        rateProgramIDs: ['abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce'],
        rateCertificationName:
            'MCR-MN-0003-SNBC-RATE-20220201-20230201-CERTIFICATION-20220202',
        rateDateStart: new Date('02/01/2022'),
        rateDateEnd: new Date('02/01/2023'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'DELOITTE',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    const rate3FormData: RateFormDataType = {
        rateType: 'AMENDMENT',
        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('01/02/2021'),
        rateProgramIDs: [
            'ea16a6c0-5fc6-4df8-adac-c627e76660ab',
            'd95394e5-44d1-45df-8151-1cc1ee66f100',
        ],
        rateCertificationName:
            'MCR-MN-0003-MSC+-PMAP-RATE-20210605-20211231-AMENDMENT-20210102',
        rateDateStart: new Date('01/01/2022'),
        rateDateEnd: new Date('01/01/2023'),
        amendmentEffectiveDateStart: new Date('06/05/2021'),
        amendmentEffectiveDateEnd: new Date('12/31/2021'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'STATE_IN_HOUSE',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    sub.packageSubmissions[0].rateRevisions = [
        {
            ...sub.packageSubmissions[0].rateRevisions[0],
            formData: rate1FormData,
        },
        {
            ...sub.packageSubmissions[0].rateRevisions[0],
            formData: rate2FormData,
        },
        {
            ...sub.packageSubmissions[0].rateRevisions[0],
            formData: rate3FormData,
        },
    ]

    const defaultStatePrograms = mockMNState().programs

    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Submission type: Contract action and rate certification'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Rating period: 01/01/2021 to 01/01/2022'
            ),
        })
    )

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Contract effective dates: 01/01/2021 to 01/01/2025'
            ),
        })
    )
    //Expect only have 3 rate names using regex to match name pattern specific to rate names.
    expect(
        template.bodyText?.match(
            /-RATE-[\d]{8}-[\d]{8}-(?:CERTIFICATION|AMENDMENT)-[\d]{8}/g
        )?.length
    ).toBe(3)
    //First Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.packageSubmissions[0].rateRevisions[0].formData
                    .rateCertificationName!
            ),
        })
    )
    //Second Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.packageSubmissions[0].rateRevisions[0].formData
                    .rateCertificationName!
            ),
        })
    )
    //Third Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.packageSubmissions[0].rateRevisions[0].formData
                    .rateCertificationName!
            ),
        })
    )
})

test('includes expected data summary for a rate amendment submission State email', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractDateStart =
        new Date('01/01/2021')
    sub.packageSubmissions[0].contractRevision.formData.contractDateEnd =
        new Date('01/01/2025')
    sub.packageSubmissions[0].rateRevisions[0].formData = {
        rateType: 'AMENDMENT',

        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('10/19/2022'),
        rateProgramIDs: ['abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce'],
        rateCertificationName:
            'MCR-MN-0003-SNBC-RATE-20210605-20211231-AMENDMENT-20221019',
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
        amendmentEffectiveDateStart: new Date('06/05/2021'),
        amendmentEffectiveDateEnd: new Date('12/31/2021'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'MERCER',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    const statePrograms = mockMNState().programs

    const template = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        statePrograms
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Submission type: Contract action and rate certification'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Rate amendment effective dates: 06/05/2021 to 12/31/2021'
            ),
        })
    )
    //Expect only have 1 rate names using regex to match name pattern specific to rate names.
    expect(
        template.bodyText?.match(
            /-RATE-[\d]{8}-[\d]{8}-(?:CERTIFICATION|AMENDMENT)-[\d]{8}/g
        )?.length
    ).toBe(1)
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.packageSubmissions[0].rateRevisions[0].formData
                    .rateCertificationName!
            ),
        })
    )
})

test('renders overall email for a new package with a rate amendment as expected', async () => {
    const sub: ContractType = mockContract()
    sub.packageSubmissions[0].contractRevision.formData.contractDateStart =
        new Date('01/01/2021')
    sub.packageSubmissions[0].contractRevision.formData.contractDateEnd =
        new Date('01/01/2025')
    sub.packageSubmissions[0].rateRevisions[0].formData = {
        rateType: 'AMENDMENT',
        rateDocuments: [
            {
                s3URL: 's3://bucketname/key/test1',
                name: 'foo',
                sha256: 'fakesha',
            },
        ],
        supportingDocuments: [],
        rateDateCertified: new Date('01/02/2021'),
        rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
        rateCertificationName:
            'MCR-MN-0003-MSHO-RATE-20210605-20211231-AMENDMENT-20210102',
        amendmentEffectiveDateStart: new Date('06/05/2021'),
        amendmentEffectiveDateEnd: new Date('12/31/2021'),
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
        certifyingActuaryContacts: [
            {
                actuarialFirm: 'MERCER',
                name: 'Actuary Contact 1',
                titleRole: 'Test Actuary Contact 1',
                email: 'actuarycontact1@example.com',
            },
        ],
        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
        packagesWithSharedRateCerts: [],
    }
    const defaultStatePrograms = mockMNState().programs
    const result = await newContractStateEmail(
        sub,
        defaultSubmitters,
        testEmailConfig(),
        defaultStatePrograms
    )

    if (result instanceof Error) {
        console.error(result)
        return
    }

    expect(result.bodyHTML).toMatchSnapshot()
})
