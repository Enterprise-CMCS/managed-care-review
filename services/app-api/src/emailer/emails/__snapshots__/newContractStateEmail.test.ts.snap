// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders overall email for a new package with a rate amendment as expected 1`] = `
"MCR-MN-0004-SNBC was successfully submitted.
<br /><br />
<b>Submission type</b>: Contract action and rate certification<br />
<b>Contract amendment effective dates</b>: 01/01/2021 to 01/01/2025<br />
<br />
<b>Rate name</b>: MCR-MN-0003-MSHO-RATE-20210605-20211231-AMENDMENT-20210102<br />
<b>Rate amendment effective dates</b>: 06/05/2021 to 12/31/2021<br />
<br />
<b>Submission description</b>: A real submission<br />
<br />
<a href="http://localhost/submissions/health-plan/test-contract-123">View submission</a>
<br /><br />
If you need to make any changes, please contact CMS.
<br /><br />

<div>What comes next:</div>
<ol>
    <li>
        <strong>Check for completeness:</strong> CMS will review all documentation submitted to ensure all required materials were received.
    </li>
    <li>
        <strong>CMS review:</strong> Your submission will be reviewed by CMS for adherence to federal regulations.
    </li>
    <li>
        <strong>Questions:</strong> You may receive questions via email from CMS as they conduct their review.
    </li>
    <li>
        <strong>Decision:</strong> Once all questions have been addressed, CMS will contact you with their final recommendation.

    </li>
</ol>
<br />
<div>If you need assistance or to make changes to your submission:</div>
<ul>
    <li>
        For assistance with programmatic, contractual, or operational issues, please reach out to mcog@example.com and/or your CMS primary contact.
    </li>
    <li>
        For assistance on policy and actuarial issues, please reach out to rates@example.com.
    </li>
    <li>
        For issues related to MC-Review or all other inquiries, please reach out to MC_Review_HelpDesk@example.com.
    </li>
</ul>
"
`;
